src/core/team-core.js

```diff
@@ -1,24 +1,98 @@
-const {knex} = require('../util/database').connect();
 import _ from 'lodash';
+import BPromise from 'bluebird';
+const {knex} = require('../util/database').connect();
+const redisClient = require('../util/redis').connect().client;
 import {deepChangeKeyCase} from '../util';
 
+
+const KEY_TEAMS = 'teams';
+const KEY_IS_STALE = 'teamsAreStale';
+const KEY_IS_UPDATING = 'teamsAreUpdating';
+const CACHE_TTL = process.env.TEAMS_CACHE_TTL ||  60 * 10; // in 's'
+let init;
+
+function initialize() {
+  init = _isStale().then(stale => stale ? _updateCache() : BPromise.resolve());
+}
+
+function _getField(cityId) {
+  return `city_${ !cityId ? 'all' : cityId }`;
+}
+
 function getTeams(opts) {
-  const isBanned = opts.client && !!opts.client.isBanned;
+  _isStale().then(stale => {
+    // Incur a cache refresh if necessary
+    if (stale) _updateCache();
+  });
+
+  // Do not await for update to complete as it may take some time.
+  return BPromise.all([init]).then(() =>
+    redisClient.hgetAsync(KEY_TEAMS, _getField(opts.city))
+      .then(result =>JSON.parse(result))
+  );
+}
+
+function _isStale() {
+  return redisClient.getAsync(KEY_IS_STALE).then(stale => !(stale === 'false'));
+}
+
+function _isUpdating() {
+  return redisClient
+    .getAsync(KEY_IS_UPDATING)
+    .then(isUpdating => isUpdating == 'true');
+}
+
+/**
+ * Updates the cache if no update is in progress. Otherwise does nothing.
+ *
+ * To note is that checking for on going update is best effort only. Overlapping
+ * updates are possible, but should be very uncommon and not impact the
+ * generated cache in a negative way.
+ **/
+function _updateCache() {
+  const update = redisClient.setAsync(KEY_IS_UPDATING, 'true')
+    .then(() => knex('cities').select('*')
+      // Cache per city team rankings
+      .then(cities =>
+        BPromise.map(cities, city =>
+          _getTeams(city.id).then(teams =>
+            redisClient.hmsetAsync(KEY_TEAMS, _getField(city.id), JSON.stringify(teams))
+          )
+        )
+      )
+      // Cache nation wide team rankings
+      .then(() =>
+        _getTeams().then(teams =>
+            redisClient.hmsetAsync(KEY_TEAMS, _getField(), JSON.stringify(teams))
+        )
+      )
+      .then(() => redisClient.msetAsync(KEY_IS_STALE, 'false', KEY_IS_UPDATING, 'false'))
+      .then(() => redisClient.expire(KEY_IS_STALE, CACHE_TTL))
+    );
+
+  return _isUpdating().then(isUpdating => isUpdating ? BPromise.resolve() : update);
```

Comment:
`update` should actually be a function. Now it performs the cache update every single time this method is called even tho the update is already ongoing. This is because the promise is initiated even tho it's not waited. The fix is to change `update` to be function:

```js
const update = () => 
...
return _isUpdating().then(isUpdating => isUpdating ? BPromise.resolve() : update());
```